### ISSUE TITLE
Coverage library returns incorrect info for hash literal constants

### ISSUE THREAD

**Issue by brasic** [jruby/jruby#7859]
I noticed missing test coverage for jruby in a (private) library I maintain and it appears that JRuby's `Coverage` implementation returns different results from MRI's in several ways. Some of these are arguably arbitrary but others seem clearly wrong.  For example, MRI lists the lines where a multiline hash constant is defined as `nil` for the constant itself (coverage disabled) and `1` for the first line of the definition. This is consistent for both constants made of pure literals and constants that require some code execution to construct. JRuby demonstrates a different behavior, and is also not consistent between the two types of hash literals.  Here's an example extracted from the reproduction script included below:

`nil`/`1` coverage lines flipped from MRI:
```
$ diff -y r-lit j-lit -W60
ruby 3.1.2p20 (2022-04-12 r  |  jruby 9.4.2.0 (3.1.0) 2023-
nil  |   LITERAL_ONLY = {    |  1    |   LITERAL_ONLY = {
1    |     "1" => "one",     |  nil  |     "1" => "one",
nil  |     "2" => "two"         nil  |     "2" => "two"
nil  |   }                      nil  |   }
```

odd difference for hash literals with non-literal keys: the line defining the constant itself as well as each value is listed as 0 coverage, even though those lines were "executed" when the constant was defined. Also, no access of the hash seems to be able to cause those 0s to increment. This seems like a clear bug. 

```
$ diff -y r-dynlit j-dynlit -W60
ruby 3.1.2p20 (2022-04-12 r  |  jruby 9.4.2.0 (3.1.0) 2023-
nil  |   COMPUTED = {        |  0    |   COMPUTED = {
1    |     A.name => A,      |  0    |     A.name => A,
nil  |     B.name => B,      |  0    |     B.name => B,
nil  |   }                      nil  |   }
```

**Environment Information**
  
* `jruby 9.4.2.0 (3.1.0) 2023-03-08 90d2913fda`
* coverage mode set using `debug.fullTrace=true` in `~/.jrubyrc`
* Does not look platform dependent, tested with Darwin Kernel Version 22.5.0 arm64 and ubuntu 22.04 amd64.
* No gems, just stdlib `coverage`.

**Expected Behavior**

JRuby matches MRI Coverage library semantics. See above for detail

**Actual Behavior**
constant lines are apparently unreachable by coverage metrics.

Here is a reproduction script:

```ruby
content = <<~EOF
  class Foo
    def initialize(name)
      @name = name
    end

    attr_reader :name

    A = new("ay")
    B = new("bee")

    LITERAL_ONLY = {
      "1" => "one",
      "2" => "two"
    }

    COMPUTED = {
      A.name => A,
      B.name => B,
    }
  end
EOF
require 'coverage'
Coverage.start
File.write("demo.rb", content)
require_relative 'demo'
p RUBY_DESCRIPTION
coverage = Coverage.result.values.first
content.lines.each_with_index do |line, i|
  print "#{coverage[i].inspect.ljust(4)} | #{line}"
end
```

Output on MRI 3.1.2:
```
"ruby 3.1.2p20 (2022-04-12 revision 4491bb740a) [arm64-darwin22]"
1    | class Foo
1    |   def initialize(name)
2    |     @name = name
nil  |   end
nil  | 
1    |   attr_reader :name
nil  | 
1    |   A = new("ay")
1    |   B = new("bee")
nil  | 
nil  |   LITERAL_ONLY = {
1    |     "1" => "one",
nil  |     "2" => "two"
nil  |   }
nil  | 
nil  |   COMPUTED = {
1    |     A.name => A,
nil  |     B.name => B,
nil  |   }
nil  | end
```

Output on JRuby 9.4.2.0:
```
"jruby 9.4.2.0 (3.1.0) 2023-03-08 90d2913fda OpenJDK 64-Bit Server VM 17.0.7+7 on 17.0.7+7 +jit [arm64-darwin]"
1    | class Foo
1    |   def initialize(name)
2    |     @name = name
nil  |   end
nil  | 
1    |   attr_reader :name
nil  | 
1    |   A = new("ay")
1    |   B = new("bee")
nil  | 
1    |   LITERAL_ONLY = {
nil  |     "1" => "one",
nil  |     "2" => "two"
nil  |   }
nil  | 
0    |   COMPUTED = {
0    |     A.name => A,
0    |     B.name => B,
nil  |   }
nil  | end
```

**Comment by enebo** [comment]
It appears we mark COMPUTED as 0 since we expect to see a line number instr.  With a method call on next line for some reason the COMPUTED line instr does not happen.  Then A.name and B.name both are calls and are marked as newline so we emit line number instrs.  However, the initial coverage map shows those two lines as -1 so when the line number happens it increments those two lines to 0.

Original starting lines for coverage happen building the AST and then we make line number instrs when we translate the AST to IR.   So two layers are not interacting well here.

**Comment by enebo** [comment]
I caused this with https://github.com/jruby/jruby/issues/4154 . newline_node() in RubyParserBase will set coverage lines to 0 vs -1.  I added a setNewLine to all CallNodes but I do not update the coverage map.  This explains the two 0 in the hash (-1  + 1 = 0).  Personally I do not find an issue with marking those calls as 1 for the calls due to the example I gave two comments up but I think I will revisit 4154 and see if we can just match MRI better (Although it will end up putting a 1 on CONSTANT and not the line after it.

**Comment by enebo** [comment]
@brasic as explained above and in the PR I do not plan on fixing the line marking the hash because it appears MRI would fix it if they could figure out a nice way to do it.  Plus any coverage using tools currently probably have logic to account for this difference (if they need to) so I think it may break someone?

The fix for the 0s is explained in the PR itself.

**Comment by brasic** [comment]
@enebo I think that makes sense. The coverage tool I’m using is https://github.com/simplecov-ruby/simplecov. I’ll verify the fix in https://github.com/jruby/jruby/pull/7863 and take up the other discrepancies with simplecov if necessary.

**Comment by enebo** [comment]
Not sure what I was looking at but there was other fallout from this fix.  Reverting until I address those new failures.

**Comment by enebo** [comment]
This is becoming a dumping ground of trying to fix what was #4154 but that issue was what caused the 0s of this issue.  The reason I think I did not fix this the way I just attempted is because the order of examining nodes while building them is convoluted.  Having a dual condition for lines which are needed for coverage/profiling and for extra lines which are needed for backtraces on errors is fine....

but... 

```determineIfWeNeedLineNumber``` is called at beginning of ```buildOperand``` but the actual line number instrs are added at the front of ```addInstr```.  The weird part is we get out of order determines but in order addInstr.  Consider:

```ruby
1.upto(0) do
end
  .
  foo(
    1,
    2,
  )
```

The AST is:
```text
RootNode line: 3
  CallNode*:foo line: 3
    CallNode:upto line: 0
      FixnumNode line: 0, long: 1
      ArrayNode line: 0
        FixnumNode line: 0, long: 0
      IterNode line: 0
        ArgsNode line: 1
          ListNode line: 1
          ListNode line: 1
          ListNode line: 1
          ListNode line: 1
    ArrayNode line: 4
      FixnumNode line: 4, long: 1
      FixnumNode line: 5, long: 2
, null
```

determine is called on foo first then called on upto second but addinstr is called on upto first and foo second.   The field state is about foo but we look at upto.

One solution would be to make addInstr know the Node it is coming from but that is a massive change.
